Make internal (core) errors show up in the Qt client.

zander opened the pull request:
Serious warnings wrt to state of the chain were only printed to the logs, with this change these issues are also shown in the Qt Gui..

zander committed:
Make internal (core) errors show up in the Qt client.

laanwj commented:
Duplicate of #7460?

laanwj commented:
Ah, you're trying to submit to two branches. Please don't open every pull for every branch. I'll cherry-pick to the 0.12 branch as appropriate.

laanwj commented:
Concept ACK.
My only concern is that these warnings will never go away (no equiv CT_DELETED).

zander commented:
Integrated feedback and made the diff minimal.

zander committed:
Scope it

LOCK is a wrapper around boost::recursive_mutex, but it can't hurt
to unlock before we pass it on.

MarcoFalke commented on src/alert.cpp:
```diff
@@ -245,8 +247,20 @@ bool CAlert::ProcessAlert(const std::vector<unsigned char>& alertKey, bool fThre
 
-void
-CAlert::Notify(const std::string& strMessage, bool fThread)
+// static
```
Nit: What is the meaning of this comment?

zander replied:
The comment is a habit I have to indicate that the method just below is static. Something that is in the signature in the header, but not possible to have in any other way in the cpp file.

jtimon commented:
Concept ACK

jtimon commented on src/alert.h:
```diff
@@ -102,7 +102,7 @@ class CAlert : public CUnsignedAlert
     bool CheckSignature(const std::vector<unsigned char>& alertKey) const;
     bool ProcessAlert(const std::vector<unsigned char>& alertKey, bool fThread = true); // fThread means run -alertnotify in a free-running thread
-    static void Notify(const std::string& strMessage, bool fThread);
+    static void NotifyInternal(const std::string& strMessage);
```
Wouldn't "NotifyExternal" be a more appropriate name that it is exposed while the static function is only used internally?

zander replied:
I'm not sure what you mean, the difference between the Notify and the Notify Internal is that the latter also sends a message to the uiClient. Which is internal, not external...

jtimon replied:
Never mind, somehow I was thinking about it being internal to alert.o...

jtimon commented on src/alert.cpp:
```diff
@@ -27,6 +27,8 @@ using namespace std;
 map<uint256, CAlert> mapAlerts;
 CCriticalSection cs_mapAlerts;
 
+static void Notify(const std::string& strMessage, bool fThread);
```
This declaration could be saved if the the definition of the newly created method was moved below the definition of the static function.

zander replied:
Actually, there are various usages of the method Notify throughout the file, not just in the method I added.

jtimon replied:
You are right, sorry.

gavinandresen commented on src/alert.cpp:
```diff
@@ -245,8 +247,20 @@ bool CAlert::ProcessAlert(const std::vector<unsigned char>& alertKey, bool fThre
+    static CAlert internalAlert;
+    internalAlert.vchMsg = std::vector<unsigned char>(strMessage.begin(), strMessage.end());
+    uint256 zero;
+    LOCK(cs_mapAlerts);
```
It would be better to hold the cs_mapAlerts lock for as little time as possible, so good to add a local scope like this:
```
{
    LOCK(cs_mapAlerts);
    mapAlerts[zero] = internalAlert;
}
```

zander commented:
To protect users running the GUI (which really means the wallet in most cases) I would like this to end up in the upcoming release.

Can you please make sure it gets included?

laanwj commented:
@zander: 
- Review-wise, it first at least needs a tested ACK.

Whether it can be included in 0.12.0 depends on whether there needs to be another Release Candidate. It looks like rc3 may pretty well be the last one (no new critical issues have come up), so it will be relabeled as final this week. This means no further code can be included. It will make it into 0.12.1 of course.

Also I've heard some discussion that the threshold for some of the warning messages needs to be bumped (@gmaxwell); they will too easily trigger, especially in the current state of affairs with people playing with forks on the network. That needs to be done in the same release, otherwise adding this to the GUI is going to result in too many false positive complaints.

jonasschnelli commented:
Hmm... not sure about that.
Just gave this some testing.
In regtest, the `nBlocks > BLOCKS_EXPECTED` check will very likely be positive, so, this is a good way to test the alert system.

Main problem is probably that the GUI alert system can only display the most recent alert and there is no way of removing the alert.

IMO there should be a GUI alert *log* and a way of "accepting" (hiding) an alert. Something like OSX Notification Center (or iOS / Android notification mechanism).

zander commented:
> Main problem is probably that the GUI alert system can only display the most recent alert and there is no way of removing the alert.
> IMO there should be a GUI alert log and a way of "accepting" (hiding) an alert. Something like OSX Notification Center (or iOS / Android notification mechanism). 

I do think that this is a future feature that would make a lot of sense to add. But such an extension to this rather simple, yet essential feature would likely make it unavailable to the 0.12 branch.

 I would be happy if you could give the credit for what this patch actually tries to do; which is to bubble up essential networking errors to the UI. Most people running the Qt client will not look at the log file, and so they will miss that they are moved off of the net due to a future hardfork or similar.

The goal is to get this into the first release.

jonasschnelli commented:
> The goal is to get this into the first release.

0.12 final has been tagged (binaries soon).
I don't see this as a urgent feature that would allow dispatching another RC round (which is to late anyway now).

IMO adding log-based-alerts to the GUI without a actually "log" (only the most recent one without a timestamp), can make things worse. If we are going to change this, it should also respect time and multiple possible alerts.

zander commented:
> I don't see this as a urgent feature

Having the GUI show you when your client stopped accepting blocks hours ago (for instance) is not an urgent feature?

I welcome anyone wanting to make this better, however you want, as long as the end user does indeed get informed when his software is too old to be able to participate on the network.

jonasschnelli commented:
> Having the GUI show you when your client stopped accepting blocks hours ago (for instance) is not an urgent feature?

A such feature would be very nice. Although, compared to other issues, this is not a reason add a quick-fix with conceptual problems (map log-alert-system to only-display-most-recent alert) and doing other RC-round or the whole bitcoin-core packaged.

We should wait until we have a proper solution and aim this for 0.13 and (maybe) for a BP to 0.12.1.

zander commented:
Well, I made this request 17 days ago, if you can't get it into 0.12.0, then what about 0.12.1 ?

Please write a tested-ack confirming that it works (which you wrote above you did), so it can be merged at least.

jonasschnelli commented:
> Well, I made this request 17 days ago, if you can't get it into 0.12.0, then what about 0.12.1 ?

Please respect the way how Core does do quality assurance. Most PR take much longer than 17 days until they get merged. Time does not matter in this case. We need reviews.

> Please write a tested-ack confirming that it works (which you wrote above you did), so it can be merged at least.

I have tested It and I don't agree it would be wise to merge. As said: this maps a log based alert system to a single, non-timestamped alert display. IMO this is not a solution we should merge.
Example: If a user faces network problems (close 8333p, etc.), he would see `WARNING: check your network connection, %d blocks received in the last %d hours (%d expected)` in the GUI, he might resolve the problem then, but the alert would never disappear (unless he restarts BitcoinQt). 

At least this PR should display the *age* (timestamp calculation) of the most recent alert and a way of hiding alerts when the condition are fulfilled (solved network issues, etc.), than I could agree that this is a step towards correct `CAlert` display in GUI

zander commented:
> Example: If a user faces network problems (close 8333p, etc.), he would see WARNING: check your network connection, %d blocks received in the last %d hours (%d expected) in the GUI, he might resolve the problem then, 

Good enough example, the current situation is that the user never sees this warning. In your example he has applied this patch, sees the problem and can fix it.  
Without this patch, he won't fix it...

Again, if its not perfect, then refine it, make it better for a future release.  Don't stop real users from seeing real problems in the first available release.

jonasschnelli commented:
> Without this patch, he won't fix it...

There is still the debug.log (there is a "open" button from the debug window).

> Again, if its not perfect, then refine it, make it better for a future release. Don't stop real users from seeing real problems in the first available release.

I just try to avoid a fix that introduces new issues.

What if we ...

* ...would add a timestamp (displays in age in m/h/d) at the top-left?
* ...Maybe add a (X) close symbol at the top right.
* ...(And a way to scroll down in multiple alerts are there)

gavinandresen commented:
Seeing an error/alert is strictly better than not.

Tested ACK from me.

jonasschnelli commented:
Did some additional testing.
If a large chain fork was reported to the GUI, the error gets overwritten by `WARNING: check your network connection, 0 blocks received in the last 4 hours (24 expected)` after 4 hours. There is a high chance of missing the error (4h time window) and confusing the user with "network troubles".

Working on a alternative solution now.

jonasschnelli commented on src/alert.cpp:
```diff
@@ -245,8 +247,22 @@ bool CAlert::ProcessAlert(const std::vector<unsigned char>& alertKey, bool fThre
+{
+    static CAlert internalAlert;
+    internalAlert.vchMsg = std::vector<unsigned char>(strMessage.begin(), strMessage.end());
+    uint256 zero;
```
I guess you need at least a `internalAlert.nExpiration` time here.

jonasschnelli commented on src/alert.cpp:
```diff
@@ -245,8 +247,22 @@ bool CAlert::ProcessAlert(const std::vector<unsigned char>& alertKey, bool fThre
+void CAlert::NotifyInternal(const std::string& strMessage)
+{
+    static CAlert internalAlert;
+    internalAlert.vchMsg = std::vector<unsigned char>(strMessage.begin(), strMessage.end());
```
IMO you also should set `internalAlert.strStatusBar`. (vchMsg is the raw data of a network alert that comes with a signature).

zander replied:
Why? There is no change at all in the user experience...

jonasschnelli replied:
`internalAlert.vchMsg` will never be shown in your case. What's the reason in instantiating a CAlert? IMO it's completely unused.

jonasschnelli commented on src/alert.cpp:
```diff
@@ -245,8 +247,22 @@ bool CAlert::ProcessAlert(const std::vector<unsigned char>& alertKey, bool fThre
+    uint256 zero;
+    {
+        LOCK(cs_mapAlerts);
+        mapAlerts[zero] = internalAlert;
```
Does this includes a risk that the alert gets relayed to other peers?

laanwj commented:
Closing in favor of #7579

MarcoFalke commented:
Pull is closed, removing 'Needs backport'
